xmtp_api/src/buffered_stream.rs

```diff
@@ -0,0 +1,190 @@
+use futures::stream::Stream;
+use std::pin::Pin;
+use std::task::{Context, Poll};
+use tokio::sync::mpsc::{unbounded_channel, UnboundedReceiver};
+use tokio::task::JoinHandle;
+
+/// BufferedStream wraps an existing stream and buffers all messages in memory
+/// to prevent backpressure to the underlying stream.
+pub struct BufferedStream<T, E> {
+    receiver: UnboundedReceiver<Result<T, E>>,
+    _handle: JoinHandle<()>,
+}
+
+impl<T, E> BufferedStream<T, E>
+where
+    T: Send + 'static,
+    E: Send + 'static,
+{
+    /// Creates a new BufferedStream from an existing stream
+    pub fn new<S>(mut stream: S) -> Self
+    where
+        S: Stream<Item = Result<T, E>> + Send + Unpin + 'static,
```

Comment:
The Unpin will be problematic in webassembly since the Reqwest stream does not implement `Unpin`. If you use `pin_project_lite` around your BufferedStreamType (using `#[pin]` for your receiver type) you will be able to get the references needed with `this.project().receiver` to call poll_next.